File: frontend/src/utils/artifactPreview.ts
import type { Artifact } from "../types";

export interface ArtifactPreviewInfo {
  kind: string;
  label: string;
  shortLabel: string;
  extension: string | null;
  mediaType: string | null;
  rawUrl: string | null;
  downloadUrl: string | null;
  canInline: boolean;
  isBinary: boolean;
}

const imageExtensions = new Set([".png", ".jpg", ".jpeg", ".gif", ".webp", ".svg", ".bmp", ".ico"]);
const markdownExtensions = new Set([".md", ".markdown", ".mdx"]);
const presentationExtensions = new Set([".ppt", ".pptx", ".key"]);
const wordExtensions = new Set([".doc", ".docx", ".rtf"]);
const spreadsheetExtensions = new Set([".xls", ".xlsx", ".csv", ".tsv"]);

export function getArtifactPreviewInfo(artifact: Artifact): ArtifactPreviewInfo {
  const extension = normalizeExtension(artifact.fileExtension) ?? extensionFromPath(artifact.filePath ?? artifact.title);
  const rawKind = artifact.previewKind ?? inferPreviewKind(artifact, extension);
  const kind = normalizeKind(rawKind);
  const label = artifact.previewLabel ?? labelForKind(kind, artifact.type);
  return {
    kind,
    label,
    shortLabel: shortLabelForKind(kind, artifact.type),
    extension,
    mediaType: artifact.mediaType ?? mediaTypeForKind(kind, extension),
    rawUrl: artifact.rawUrl ?? null,
    downloadUrl: artifact.downloadUrl ?? artifact.rawUrl ?? null,
    canInline: artifact.canInlinePreview ?? canInlineKind(kind),
    isBinary: artifact.isBinary ?? isBinaryKind(kind),
  };
}

export function artifactDisplayTitle(artifact: Artifact): string {
  return artifact.title || artifact.filePath || "产物";
}

export function isMetadataOnlyContent(content: string): boolean {
  try {
    const parsed = JSON.parse(content) as unknown;
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return false;
    const data = parsed as Record<string, unknown>;
    return typeof data.path === "string" && typeof data.previewKind === "string";
  } catch {
    return false;
  }
}

function inferPreviewKind(artifact: Artifact, extension: string | null): string {
  if (artifact.type === "code_diff") return "diff";
  if (artifact.type === "file_tree") return "file_tree";
  if (artifact.type === "web_preview") return "html";
  if (extension && markdownExtensions.has(extension)) return "markdown";
  if (extension === ".pdf") return "pdf";
  if (extension && imageExtensions.has(extension)) return "image";
  if (extension && presentationExtensions.has(extension)) return "presentation";
  if (extension && wordExtensions.has(extension)) return "word";
  if (extension && spreadsheetExtensions.has(extension)) return "spreadsheet";
  if (/^#{1,6}\s+\S+/m.test(artifact.content)) return "markdown";
  return "text";
}

function normalizeKind(kind: string): string {
  if (kind === "md") return "markdown";
  if (kind === "docx" || kind === "doc") return "word";
  if (kind === "ppt" || kind === "pptx") return "presentation";
  if (kind === "xls" || kind === "xlsx") return "spreadsheet";
  return kind;
}

function labelForKind(kind: string, type: Artifact["type"]): string {
  if (kind === "diff") return "代码变更";
  if (kind === "file_tree") return "文件变更";
  if (kind === "html") return type === "web_preview" ? "网页预览" : "HTML";
  if (kind === "markdown") return "Markdown 文档";
  if (kind === "pdf") return "PDF 文档";
  if (kind === "image") return "图片预览";
  if (kind === "presentation") return "演示文稿";
  if (kind === "word") return "Word 文档";
  if (kind === "spreadsheet") return "表格文件";
  return "文档";
}

function shortLabelForKind(kind: string, type: Artifact["type"]): string {
  if (kind === "diff") return "Diff";
  if (kind === "file_tree") return "文件";
  if (kind === "html") return type === "web_preview" ? "网页" : "HTML";
  if (kind === "markdown") return "MD";
  if (kind === "pdf") return "PDF";
  if (kind === "image") return "图片";
  if (kind === "presentation") return "PPT";
  if (kind === "word") return "DOCX";
  if (kind === "spreadsheet") return "表格";
  return "文档";
}

function canInlineKind(kind: string): boolean {
  return ["diff", "file_tree", "html", "markdown", "pdf", "image", "text", "spreadsheet"].includes(kind);
}

function isBinaryKind(kind: string): boolean {
  return ["pdf", "image", "presentation", "word"].includes(kind);
}

function mediaTypeForKind(kind: string, extension: string | null): string | null {
  if (kind === "html") return "text/html";
  if (kind === "markdown") return "text/markdown";
  if (kind === "pdf") return "application/pdf";
  if (kind === "image") return extension === ".svg" ? "image/svg+xml" : "image/*";
  if (kind === "diff") return "text/x-diff";
  if (kind === "file_tree") return "application/json";
  if (kind === "spreadsheet" && extension === ".csv") return "text/csv";
  return null;
}

function extensionFromPath(path: string | null | undefined): string | null {
  if (!path) return null;
  const match = /\.([A-Za-z0-9]+)$/.exec(path.trim());
  return match ? `.${match[1].toLowerCase()}` : null;
}

function normalizeExtension(extension: string | null | undefined): string | null {
  if (!extension) return null;
  const clean = extension.trim().toLowerCase();
  return clean.startsWith(".") ? clean : `.${clean}`;
}

File: backend/app/services/artifact_preview.py
"""Artifact 预览格式推断。"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import quote


IMAGE_MEDIA_TYPES = {
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".webp": "image/webp",
    ".svg": "image/svg+xml",
    ".bmp": "image/bmp",
    ".ico": "image/x-icon",
}

MARKDOWN_EXTENSIONS = {".md", ".markdown", ".mdx"}
PDF_EXTENSIONS = {".pdf"}
PRESENTATION_EXTENSIONS = {".ppt", ".pptx", ".key"}
WORD_EXTENSIONS = {".doc", ".docx", ".rtf"}
SPREADSHEET_EXTENSIONS = {".xls", ".xlsx", ".csv", ".tsv"}
TEXT_EXTENSIONS = {
    ".txt",
    ".log",
    ".json",
    ".jsonl",
    ".yaml",
    ".yml",
    ".toml",
    ".xml",
    ".css",
    ".js",
    ".jsx",
    ".ts",
    ".tsx",
    ".py",
    ".sql",
    ".sh",
    ".ps1",
}


@dataclass(frozen=True)
class ArtifactPreviewInfo:
    kind: str
    label: str
    media_type: str | None
    extension: str | None
    can_inline: bool
    is_binary: bool

    def to_payload(self, artifact_id: str, has_file: bool) -> dict[str, str | bool | None]:
        raw_url = f"/api/artifacts/{quote(artifact_id, safe='')}/raw" if has_file else None
        return {
            "previewKind": self.kind,
            "previewLabel": self.label,
            "mediaType": self.media_type,
            "fileExtension": self.extension,
            "canInlinePreview": self.can_inline,
            "isBinary": self.is_binary,
            "rawUrl": raw_url,
            "downloadUrl": f"{raw_url}?download=true" if raw_url else None,
        }


def infer_artifact_preview(
    *,
    artifact_type: str,
    title: str | None,
    content: str | None,
    file_path: str | None,
) -> ArtifactPreviewInfo:
    extension = _extension(file_path or title)
    clean_type = (artifact_type or "").strip()
    body = content or ""

    if clean_type == "code_diff":
        return ArtifactPreviewInfo("diff", "代码 Diff", "text/x-diff", extension, True, False)
    if clean_type == "file_tree":
        return ArtifactPreviewInfo("file_tree", "文件变更", "application/json", extension, True, False)
    if clean_type == "web_preview":
        return ArtifactPreviewInfo("html", "网页预览", "text/html", extension or ".html", True, False)

    if extension in MARKDOWN_EXTENSIONS:
        return ArtifactPreviewInfo("markdown", "Markdown 文档", "text/markdown", extension, True, False)
    if extension in PDF_EXTENSIONS:
        return ArtifactPreviewInfo("pdf", "PDF 文档", "application/pdf", extension, True, True)
    if extension in IMAGE_MEDIA_TYPES:
        return ArtifactPreviewInfo("image", "图片", IMAGE_MEDIA_TYPES[extension], extension, True, True)
    if extension in PRESENTATION_EXTENSIONS:
        return ArtifactPreviewInfo(
            "presentation",
            "演示文稿",
            _office_media_type(extension),
            extension,
            False,
            True,
        )
    if extension in WORD_EXTENSIONS:
        return ArtifactPreviewInfo("word", "Word 文档", _office_media_type(extension), extension, False, True)
    if extension in SPREADSHEET_EXTENSIONS:
        media_type = "text/csv" if extension in {".csv", ".tsv"} else _office_media_type(extension)
        return ArtifactPreviewInfo("spreadsheet", "表格文件", media_type, extension, extension in {".csv", ".tsv"}, extension not in {".csv", ".tsv"})
    if extension in TEXT_EXTENSIONS:
        return ArtifactPreviewInfo("text", "文本文件", _text_media_type(extension), extension, True, False)
    if _looks_like_markdown(body):
        return ArtifactPreviewInfo("markdown", "Markdown 文档", "text/markdown", extension, True, False)
    if _looks_like_pdf(body):
        return ArtifactPreviewInfo("pdf", "PDF 文档", "application/pdf", extension, True, True)
    return ArtifactPreviewInfo("text", "文档", "text/plain", extension, True, False)


def artifact_preview_payload(artifact) -> dict[str, str | bool | None]:
    info = infer_artifact_preview(
        artifact_type=str(getattr(artifact, "type", "") or ""),
        title=str(getattr(artifact, "title", "") or ""),
        content=str(getattr(artifact, "content", "") or ""),
        file_path=getattr(artifact, "file_path", None),
    )
    return info.to_payload(str(getattr(artifact, "id")), bool(getattr(artifact, "file_path", None)))


def _extension(path_or_title: str | None) -> str | None:
    if not path_or_title:
        return None
    suffix = Path(path_or_title.strip()).suffix.lower()
    return suffix or None


def _looks_like_markdown(content: str) -> bool:
    return bool(re.search(r"^#{1,6}\s+\S+", content, re.MULTILINE))


def _looks_like_pdf(content: str) -> bool:
    return content.startswith("%PDF")


def _text_media_type(extension: str | None) -> str:
    if extension == ".json":
        return "application/json"
    if extension in {".yaml", ".yml"}:
        return "application/yaml"
    if extension == ".xml":
        return "application/xml"
    if extension == ".css":
        return "text/css"
    if extension in {".js", ".jsx", ".ts", ".tsx"}:
        return "text/javascript"
    return "text/plain"


def _office_media_type(extension: str | None) -> str:
    if extension == ".docx":
        return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
    if extension == ".pptx":
        return "application/vnd.openxmlformats-officedocument.presentationml.presentation"
    if extension == ".xlsx":
        return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
    if extension == ".doc":
        return "application/msword"
    if extension == ".ppt":
        return "application/vnd.ms-powerpoint"
    if extension == ".xls":
        return "application/vnd.ms-excel"
    if extension == ".rtf":
        return "application/rtf"
    return "application/octet-stream"
